### ISSUE TITLE
Trouble Running Browser example

### ISSUE THREAD

**Issue by ashays** [haadcode/orbit-db#113]
Wanted to play around with orbit-db for a project but was having trouble getting the browser example working (though the Node.js example did work). I'm using Mac OS X with Node v6.4.0 and npm version 3.10.8. Would appreciate any assistance or feedback about what the problem may be! Thanks a ton

Here's the output when I try to execute `npm run build:examples` after `npm install` did not error:

``` javascript
> orbit-db@0.13.2 build:examples /Users/employee/Documents/gilbert/orbit-db
> webpack --config conf/webpack.example.config.js

Hash: fc0cf69b5c90511c787d
Version: webpack 2.1.0-beta.25
Time: 17644ms
                       Asset    Size  Chunks             Chunk Names
./examples/browser/bundle.js  282 kB       0  [emitted]  main
    + 151 hidden modules

ERROR in ./~/exports-loader?IpfsApi!./~/ipfs-api/dist/index.js
Module build failed: SyntaxError: Deleting local variable in strict mode (26454:4)

  26452 |   deps = (typeof ids === 'string') ? factory.slice(2) : ids.slice(2);
  26453 |   if(nodeJS) {
> 26454 |     delete define;
        |     ^
  26455 |     return tmpDefine.apply(null, Array.prototype.slice.call(arguments, 0));
  26456 |   }
  26457 |   define = tmpDefine;

 @ ./examples/browser/index.js 3:14-63

ERROR in ./~/orbit-db-eventstore/src/EventStore.js
Module not found: Error: Can't resolve 'lodash.slice' in '/Users/employee/Documents/gilbert/orbit-db/node_modules/orbit-db-eventstore/src'
 @ ./~/orbit-db-eventstore/src/EventStore.js 37:12-35
 @ ./src/OrbitDB.js
 @ ./examples/browser/index.js

ERROR in ./~/orbit-db-eventstore/src/EventStore.js
Module not found: Error: Can't resolve 'lodash.findindex' in '/Users/employee/Documents/gilbert/orbit-db/node_modules/orbit-db-eventstore/src'
 @ ./~/orbit-db-eventstore/src/EventStore.js 39:16-43
 @ ./src/OrbitDB.js
 @ ./examples/browser/index.js

ERROR in ./~/orbit-db-counterstore/src/CounterIndex.js
Module not found: Error: Can't resolve 'crdts/src/G-Counter' in '/Users/employee/Documents/gilbert/orbit-db/node_modules/orbit-db-counterstore/src'
 @ ./~/orbit-db-counterstore/src/CounterIndex.js 13:14-44
 @ ./~/orbit-db-counterstore/src/CounterStore.js
 @ ./src/OrbitDB.js
 @ ./examples/browser/index.js

ERROR in (webpack)/~/node-libs-browser/~/buffer/index.js
Module not found: Error: Can't resolve 'base64-js' in '/Users/employee/Documents/gilbert/orbit-db/node_modules/webpack/node_modules/node-libs-browser/node_modules/buffer'
 @ (webpack)/~/node-libs-browser/~/buffer/index.js 11:13-33
 @ ./~/ipfs-log/src/log.js
 @ ./~/orbit-db-store/src/Store.js
 @ ./~/orbit-db-eventstore/src/EventStore.js
 @ ./src/OrbitDB.js
 @ ./examples/browser/index.js

ERROR in (webpack)/~/node-libs-browser/~/buffer/index.js
Module not found: Error: Can't resolve 'ieee754' in '/Users/employee/Documents/gilbert/orbit-db/node_modules/webpack/node_modules/node-libs-browser/node_modules/buffer'
 @ (webpack)/~/node-libs-browser/~/buffer/index.js 12:14-32
 @ ./~/ipfs-log/src/log.js
 @ ./~/orbit-db-store/src/Store.js
 @ ./~/orbit-db-eventstore/src/EventStore.js
 @ ./src/OrbitDB.js
 @ ./examples/browser/index.js

ERROR in (webpack)/~/node-libs-browser/~/timers-browserify/main.js
Module not found: Error: Can't resolve 'process/browser.js' in '/Users/employee/Documents/gilbert/orbit-db/node_modules/webpack/node_modules/node-libs-browser/node_modules/timers-browserify'
 @ (webpack)/~/node-libs-browser/~/timers-browserify/main.js 1:15-44
 @ ./~/bluebird/js/browser/bluebird.js
 @ ./~/ipfs-log/src/log.js
 @ ./~/orbit-db-store/src/Store.js
 @ ./~/orbit-db-eventstore/src/EventStore.js
 @ ./src/OrbitDB.js
 @ ./examples/browser/index.js

npm ERR! Darwin 14.5.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "build:examples"
npm ERR! node v6.4.0
npm ERR! npm  v3.10.8
npm ERR! code ELIFECYCLE
npm ERR! orbit-db@0.13.2 build:examples: `webpack --config conf/webpack.example.config.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the orbit-db@0.13.2 build:examples script 'webpack --config conf/webpack.example.config.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the orbit-db package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     webpack --config conf/webpack.example.config.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs orbit-db
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls orbit-db
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/employee/Documents/gilbert/orbit-db/npm-debug.log
```

**Comment by haadcode** [comment]
Hi @ashays. I did a fresh clone of orbit-db, ran npm install and npm run build:examples and it builds fine (no error). Can you try `rm -rf node_modules/`, run npm install and npm run build:examples again and see if it works.

If that doesn't work, I'm not sure what's going on. Let us know if it still errors.

**Comment by haadcode** [comment]
ping @ashays did you get it working or are you still running into issues? Let me know if this resolved your problem and we can close the issue.

**Comment by ashays** [comment]
Alright cool! Yeah that did work actually though running into something else now. I can successfully run both `npm run build:examples` and `npm run examples:browser` commands but in index.html I get the "Client request error: Failed to fetch Waiting for IPFS daemon to start..." message. Included more details in [this gist](https://gist.github.com/ashays/103d3f5175ada4631f6f5fa055f5d99e) with what messages show up in the terminal, browser, and console.

Thank you so much for the help! Very much appreciated =

**Comment by compsocial** [comment]
Hi @haadcode, one thing I'm struggling to understand is whether this can all run self-contained in the browser or not. Looking at gist that @ashays linked, there is server-side javascript that needs to run? Does the browser talk to this? 

For the project we're looking at building, self-contained browser-based IPFS would be essential (at least I believe). Thanks for your help!

**Comment by haadcode** [comment]
@compsocial it doesn't yet work fully independently in the browser, but we're working on to fix this. Meanwhile, you can start your development with js-ipfs-api using the native go-ipfs as the API for native and browser will be the same. Hope this helps.

**Comment by compsocial** [comment]
@haadcode Totally understand. Any outlook on when that might happen? Is it a deep technical problem, or right around the corner? Sorry for the hard question, but it might affect how much and when we commit to this project.

**Comment by haadcode** [comment]
@compsocial these are great questions! :) 

Short answer: it's around the corner. 

Long answer: orbit-db was actually working couple of months ago with [js-ipfs](https://github.com/ipfs/js-ipfs) (which is the stand-alone browser implementation). The reason it's not working currently is that we added (distributed) Pubsub to IPFS some time ago and we're still working to land it in js-ipfs. js-ipfs-api already has this, as does go-ipfs, and orbit-db is using that implementation currently. In terms of the API, like said, it'll be exactly the same, so once js-ipfs has everything needed, you can use orbit-db with only changing the IPFS instance you inject to Orbit, ie. instead of:
```javascript
const IPFS = require('ipfs-api')
const ipfs = new IPFS()
const orbitdb = new OrbitDB(ipfs)
```
all you have to do it:
```javascript
const IPFS = require('ipfs')
const ipfs = new IPFS()
const orbitdb = new OrbitDB(ipfs)
```

So, we have to land pubsub in js-ipfs, cleanup everything around it and orbit-db will functional as a stand-alone browser library. I don't want to give any timelines as I really don't know when exactly we can land it, but we're committed to finish it by end of this year, perhaps earlier.

Does that answer your question? Let me know if you have any further questions.

**Comment by compsocial** [comment]
@haadcode Wonderful, and fantastic explanation. We'll definitely go forward then. I appreciate the work you've done (and are doing), and it's nice to have such a pleasant open source interaction. :-)

**Comment by haadcode** [comment]
@compsocual great! Looking forward to see what you're working on, keep us posted. And if you run into any problems or have questions, don't hesitate to open an issue.

**Comment by haadcode** [comment]
@ashays coming back to your problem: van you make sure you have version 0.15.0 of orbit-db installed and try again? I believe the error you're seeing is due to an older version. Let me know if that fixes the problem.

**Comment by chriswingler** [comment]
Hi, similar problem building examples. Fresh clone.

`ERROR in ./src/OrbitDB.js
Module not found: Error: Can't resolve 'babel' in '/home/chris/workspace/orbit-db'
BREAKING CHANGE: It's no longer allowed to omit the '-loader' prefix when using loaders.
                 You need to specify 'babel-loader' instead of 'babel'.
 @ ./src/OrbitDB.js 22:20-48
 @ ./examples/browser/index.js`

**Comment by haadcode** [comment]
Hi @chriswingler I believe this is due to changes in webpack2 beta and I've seen and fixed the same errors in other repos. I'll get a fix in today, but if you're in a rush, this can be fixed by making sure you have webpack@2.1.0-beta.21 installed (instead of latest, beta.27 I believe) in the module's repo.

**Comment by haadcode** [comment]
@chriswingler this is now fixed in master and in version 0.15.2. Please let me know if you have any further problems. 

Thank you very much for reporting the bug! :)

**Comment by haadcode** [comment]
Since this was fixed in the latest release and I haven't heard back from @ashays, I'm gonna assume this is solved. If that's not the case, feel free to re-open this issue and we'll sort it out.

Thank you everyone for your contributions! These kind of issues help us know where the problems are and fix them, so reporting issues is hugely valuable. Thank you.
